### PR TITLE
[6360] Placements validator

### PR DESCRIPTION
**Pull request by defong**
### Context
Placements validator

### Changes proposed in this pull request
Add/amend the validators so that they are executed or by pass in the right order

### Guidance to review
it works

### Important business

- [ ] Does this PR introduce any PII fields that need to be overwritten or deleted in db/scripts/sanitise.sql?
- [ ] Does this PR change the database schema? If so, have you updated the config/analytics.yml file and considered whether you need to send 'import_entity' events?
- [ ] Do we need to send any updates to DQT as part of the work in this PR?
- [ ] Does this PR need an ADR?

NB: Please notify the #twd_data_insights team and ask for a review if new fields are being added to analytics.yml

### CODE REVIEW

File: app/forms/submissions/base_validator.rb [hunk 1]
```diff
@@ -33,6 +34,10 @@ def initialize(trainee:)
       @trainee = trainee
     end
 
+    def requires_placements?
+      trainee.requires_placements? && FeatureService.enabled?(:trainee_placement)
```
Comment (darokel) [inline app/forms/submissions/base_validator.rb:38]: Feels like we should just put this feature flag check directly in `trainee.requires_placements?`?

**Review by darokel — APPROVED** [review]
Just left one comment but not a blocker

File: app/forms/submissions/base_validator.rb (hunk 1 above)
Comment (defong) [reply app/forms/submissions/base_validator.rb:38]: I was thinking that we will be like removing the flags after release, maybe we defer til then?

**Comment by defong** [comment]
[Placements validator to check total number of placements etc.](https://trello.com/c/yDEzv6W1/6360-placements-validator-to-check-total-number-of-placements-etc)

File: config/locales/en.yml [hunk 2]
```diff
@@ -694,7 +694,7 @@ en:
         study_mode: *course_path
         primary_course_subjects: *course_path
         iqts_country: edit_trainee_iqts_country_path
-        placement_detail: edit_trainee_placements_details_path
+        placement_detail: trainee_placements_confirm_path
```
Comment (stevehook) [inline config/locales/en.yml:697]: When do we use the edit details path now? (Just curious)

File: app/controllers/trainees/placements/confirm_details_controller.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+      def update
+        trainee.draft? ? toggle_trainee_progress_field : @form.save!
+
+        flash[:success] = "Trainee #{flash_message_title} updated"
+
+        redirect_to(page_tracker.last_origin_page_path || trainee_path(trainee))
+      end
+    end
+  end
+end
```
Comment (stevehook) [inline app/controllers/trainees/placements/confirm_details_controller.rb:15]: :+1: makes sense to override the defaults here.

File: app/forms/placements_form.rb [hunk 4]
```diff
@@ -79,7 +81,7 @@ def missing_fields
     if valid?
       []
     else
-      "placement_detail"
+      [[:placement_detail]]
```
Comment (stevehook) [inline app/forms/placements_form.rb:84]: :+1:

File: app/forms/submissions/base_validator.rb (hunk 1 above)
Comment (stevehook) [reply app/forms/submissions/base_validator.rb:38]: My take on this is optimise for ease of removal. Fewer places we have to change the better.

File: config/locales/en.yml (hunk 2 above)
Comment (defong) [reply config/locales/en.yml:697]: It is for the blue notification on this page

image

taking the user to the edit page seems a bit wrong as they may of filled it in

image

File: app/forms/submissions/base_validator.rb (hunk 1 above)
Comment (defong) [reply app/forms/submissions/base_validator.rb:38]: https://github.com/DFE-Digital/register-trainee-teachers/pull/3734

**Pull request merged**